torch/nn/utils/prune.py

```diff
@@ -0,0 +1,1269 @@
+            n (int, float, inf, -inf, 'fro', 'nuc'): See documentation of valid
+                entries for argument p in torch.norm
+            axis (int, optional): index of the axis along which we define
+                channels to prune. Default: -1.
+        """
+        # Check range of validity of amount
+        _validate_pruning_amount_init(amount)
+        self.amount = amount
+        self.n = n
+        self.axis = axis
+
+    def compute_mask(self, t, default_mask):
+        r"""Computes and returns a mask for the input tensor `t`.
+        Starting from a base `default_mask` (which should be a mask of ones if 
+        the tensor has not been pruned yet), generate a mask to apply on top of
+        the `default_mask` by zeroing out the channels along the specified
+        axis with the lowest Ln-norm.
+
+        Args:
+            t (torch.Tensor): tensor representing the parameter to prune
+            default_mask (torch.Tensor): Base mask from previous pruning 
+                iterations, that need to be respected after the new mask is 
+                applied.  Same dims as `t`.
+
+        Returns:
+            mask (torch.Tensor): mask to apply to `t`, of same dims as `t`.
+
+        Raises:
+            IndexError: if `self.axis >= len(t.shape)`
+        """
+        # Check that tensor has structure (i.e. more than 1 dimension) such
+        # that the concept of "channels" makes sense
+        _validate_structured_pruning(t)
+        # Check that self.axis is a valid axis to index t, else raise IndexError
+        _validate_pruning_axis(t, self.axis)
+
+        # Check that the amount of channels to prune is not > than the number of
+        # channels in t along the axis to prune
+        tensor_size = t.shape[self.axis]
+        # Compute number of units to prune: amount if int,
+        # else amount * tensor_size
+        nparams_toprune = _compute_nparams_toprune(self.amount, tensor_size)
+        nparams_tokeep = tensor_size - nparams_toprune
+        # This should raise an error if the number of units to prune is larger
+        # than the number of units in the tensor
+        _validate_pruning_amount(nparams_toprune, tensor_size)
+
+        # Structured pruning prunes entire channels so we need to know the
+        # L_n norm along each channel to then find the topk based on this
+        # metric
+        norm = _compute_norm(t, self.n, self.axis)
+        # largest=True --> top k; largest=False --> bottom k
+        # Keep the largest k channels along dim=self.axis
+        topk = torch.topk(
+            norm,
+            k=nparams_tokeep,
+            largest=True,
+        )
+        # topk will have .indices and .values
+
+        # Compute binary mask by initializing it to all 0s and then filling in
+        # 1s wherever topk.indices indicates, along self.axis.
+        # mask has the same shape as tensor t
+        def make_mask(t, axis, indices):
```

Comment:
do I spot a lot of code reuse between this and the previous RandomStructuredPruning class, or did I just stare at them wrong?

Reply:
the only block that is actually identical is these few lines at the beginning of `compute_mask`:
```python
_validate_structured_pruning(t)
_validate_pruning_dim(t, self.dim)
tensor_size = t.shape[self.dim]
nparams_toprune = _compute_nparams_toprune(self.amount, tensor_size)
nparams_tokeep = tensor_size - nparams_toprune
_validate_pruning_amount(nparams_toprune, tensor_size)
```
Feels a bit premature to pull these out.